spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnGraph.scala

```diff
@@ -0,0 +1,372 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.nn.mkldnn
+
+import breeze.linalg.Axis._1
+import com.intel.analytics.bigdl.nn.Graph.ModuleNode
+import com.intel.analytics.bigdl.nn.abstractnn.{AbstractModule, Activity}
+import com.intel.analytics.bigdl.nn.tf.{ControlDependency, WithoutInput}
+import com.intel.analytics.bigdl.nn.{StaticGraph, mkldnn}
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.{LayerException, Node, T}
+import com.intel.analytics.bigdl.nn
+
+import scala.reflect.ClassTag
+
+
+class DnnGraph(
+  private val _inputs : Seq[ModuleNode[Float]],
+  private val _outputs : Seq[ModuleNode[Float]],
+  private val enableExcludeChecking: Boolean = true)
+  extends StaticGraph[Float](_inputs, _outputs, None, enableExcludeChecking) {
+  private var forwardExecution: Array[Node[AbstractModule[Activity, Activity, Float]]] = _
+  private var backwardExecution: Array[Node[AbstractModule[Activity, Activity, Float]]] = _
+  private var inputCache: Array[Activity] = _
+  private var backId2ForwardId: Array[Int] = _
+
+  @transient private var compiled: Boolean = false
+  @transient protected lazy val reorderManager = new ReorderManager()
+
+  if (enableExcludeChecking) {
+    excludeInvalidLayers(forwardExecution.map {_.element})
+  }
+
+  buildBackwardGraph()
+
+  override def updateOutput(input: Activity): Activity = {
+    // compile dnn graph according to model status
+    if (!compiled) compile()
```

Comment:
If there're a Graph `graph` which contains another Graph named `subGraph`. So when the `subGraph.compiled` is set?

Reply:
it should be very intuitive one graph contains another, right ?

Reply:
Yeah, in long term, we need to support this case. Maybe I can break down sub dnnGraph to get its inputs/outputs nodes and merge those nodes to dnnGraph when building dnnGraph.
Now, dnn layers are limited and no such case that is much necessary to do this change (maybe over design ).
So I doubt if it is worth doing now. Maybe we can add it to "todo list".